### PR TITLE
rgw: set object accounted size correctly

### PR DESCRIPTION
**Pull request by fangyuxiangGL**
sometimes, object accounted size is set wrong,
because we don't know the object size if don't resort to the compression info or manifest without object index entry.
e.g, when i use s3cmd do copy object(bucket_A/obj_A -> bucket_B/obj_B, assume the size of obj_A is 4M).
then i use s3cmd do list bucket, I got obj_B size is 512K, it is the head size apparently.

### CODE REVIEW

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -8741,6 +8741,8 @@ int RGWRados::get_obj_state_impl(RGWObjectCtx *rctx, const RGWBucketInfo& bucket
       s->manifest.set_head(bucket_info.placement_rule, obj, s->size); /* patch manifest to reflect the head we just read, some manifests might be
                                              broken due to old bugs */
       s->size = s->manifest.get_obj_size();
+      if (s->accounted_size < s->size)
```
Comment (rzarzynski) [inline src/rgw/rgw_rados.cc]: Why do we need the check? Shouldn't we always report the `accounted_size` as the result of `get_obj_size()`?

Comment (fangyuxiangGL) [inline src/rgw/rgw_rados.cc]: assume that compression is enabled, get_obj_size() will less than the original size.

Comment (cbodley) [inline src/rgw/rgw_rados.cc]: this is making sense to me now. `accounted_size` defaults to `size` above, before looking at the `RGWCompressionInfo`. if we update `size` here, that update should also be applied to `accounted_size` as well

things are tricky if -both- `accounted_size` is set from `RGWCompressionInfo` -and- `size` is set based on the manifest. while it's possible for compression to increase the object size, i think it's safe to report to smaller of the two in this unlikely case

**Comment by fangyuxiangGL** [comment]
@rzarzynski

**Comment by cbodley** [comment]
@fangyuxiangGL i haven't been able to reproduce this one. here's what i tried:
```
$ cd ~/ceph/build
$ OSD=1 MON=1 MDS=0 MGR=0 RGW=1 ../src/vstart.sh -n -d
$ bin/radosgw-admin -c ceph.conf zone placement modify --rgw-zone=default --placement-id=default-placement --compression=snappy
[restart radosgw]

$ s3cmd mb s3://BUCKET s3://BUCKET2
Bucket 's3://BUCKET/' created
Bucket 's3://BUCKET2/' created

$ s3cmd put big.iso s3://BUCKET/                                                                                                                
upload: 'big.iso' -> 's3://BUCKET/big.iso'  [part 1 of 69, 15MB] [1 of 1]
 15728640 of 15728640   100% in    2s     7.40 MB/s  done
...
upload: 'big.iso' -> 's3://BUCKET/big.iso'  [part 69 of 69, 4MB] [1 of 1]
 4194304 of 4194304   100% in    0s    97.11 MB/s  done

$ s3cmd ls s3://BUCKET/
2017-05-15 14:14 1073741824   s3://BUCKET/big.iso

$ s3cmd cp s3://BUCKET/big.iso s3://BUCKET2/
remote copy: 's3://BUCKET/big.iso' -> 's3://BUCKET2/big.iso'

$ s3cmd ls s3://BUCKET2/
2017-05-15 14:15 1073741824   s3://BUCKET2/big.iso
```
can you share your steps to reproduce this? we'll also want to create a tracker ticket for backport

**Comment by fangyuxiangGL** [comment]
@cbodley 

I think circumstance as below should be meet if you want to reproduce it. 

1. BUCKET should not be configured with compress, so it won't get the right accounted size from compression info.

2. BUCKET  and BUCKET2 should be configured with same data pool, so RGWRados::copy_obj won't return early(e.g,  copy_data is True and go into RGWRados::copy_obj_data, then return )

Frankly speaking, I found this bug in jewel, so it may be fixed in master, but I just read the code from master, i think this bug still exists.

**Comment by cbodley** [comment]
@fangyuxiangGL okay thanks, it does reproduce when compression is off

File: src/rgw/rgw_rados.cc (hunk 1 above)
Comment (cbodley) [inline src/rgw/rgw_rados.cc]: you're right that this needs to be conditional, but i'd like to be more explicit about the condition being `if (!compressed)`. can you add a flag above for that? something like:
```diff
   s->accounted_size = s->size;
 
   auto iter = s->attrset.find(RGW_ATTR_COMPRESSION);
-  if (iter != s->attrset.end()) {
+  const bool compressed = iter != s->attrset.end();
+  if (compressed) {
     // use uncompressed size for accounted_size
     try {
```

Comment (fangyuxiangGL) [inline src/rgw/rgw_rados.cc]: ok, no problem

Comment (cbodley) [inline src/rgw/rgw_rados.cc]: thanks for adding the `compressed` flag - can you update this check to say `if (!compressed)`?

**Comment by cbodley** [comment]
one of your commits is changing the Beast submodule, can you undo that part?

**Comment by fangyuxiangGL** [comment]
so faint...

it's ok now...

**Comment by fangyuxiangGL** [comment]
retest this please

**Comment by fangyuxiangGL** [comment]
@cbodley 

it's ok now..

**Comment by cbodley** [comment]
thanks @fangyuxiangGL. your initial commit has `if (s->accounted_size < s->size)` - can you please change that part to `if (!compressed)` and squash the two commits?

**Comment by fangyuxiangGL** [comment]
@cbodley 

done it.. 
is ok?

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -8710,13 +8710,15 @@ int RGWRados::get_obj_state_impl(RGWObjectCtx *rctx, const RGWBucketInfo& bucket
...
+  const bool compressed = (iter != s->attrset.end());
+  if (compressed) {
     // use uncompressed size for accounted_size
     try {
       RGWCompressionInfo info;
       auto p = iter->second.begin();
       ::decode(info, p);
-      s->accounted_size = info.orig_size;
+      if (s->accounted_size < info.orig_size)
+        s->accounted_size = info.orig_size;
```
Comment (cbodley) [inline src/rgw/rgw_rados.cc]: this change isn't needed

File: src/rgw/rgw_rados.cc (hunk 1 above)
Comment (cbodley) [inline src/rgw/rgw_rados.cc]: please change this check from `if (s->accounted_size < s->size)` to `if (!compressed)`

File: src/rgw/rgw_rados.cc (hunk 2 above)
Comment (fangyuxiangGL) [inline src/rgw/rgw_rados.cc]: sorry, I misunderstood.. will do it..

**Comment by fangyuxiangGL** [comment]
@cbodley 

done..

**Comment by cbodley** [comment]
@fangyuxiangGL thanks, looks good. i'd like to get this backported to kraken as well, could you please add this to your commit message? `Fixes: http://tracker.ceph.com/issues/20071`

**Comment by fangyuxiangGL** [comment]
@cbodley 

added

**Comment by fangyuxiangGL** [comment]
@cbodley

**Comment by cbodley** [comment]
@fangyuxiangGL thanks, i'll include in my next teuthology testing branch. hopefully today

**Comment by cbodley** [comment]
passed rgw suite: http://pulpito.ceph.com/cbodley-2017-06-06_14:33:04-rgw-wip-cbodley-testing-distro-basic-mira/

**Pull request merged**